zmq_addon.hpp

```diff
@@ -408,190 +408,115 @@ class multipart_t
-        assert(str == "Frame6");
-
-        str = multipart.popstr();
-        assert(multipart.size() == 1);
-        assert(str == "Frame7");
-
-        num = multipart.poptyp<float>();
-        assert(multipart.empty());
-        assert(num == 8.0f);
-
-        // Test message constructors and concatenation
-        multipart_t head("One", 3);
-        head.addstr("Two");
-        assert(head.size() == 2);
-
-        multipart_t tail("One-hundred");
-        tail.pushstr("Ninety-nine");
-        assert(tail.size() == 2);
-
-        multipart_t tmp(message_t("Fifty", 5));
-        assert(tmp.size() == 1);
-
-        multipart_t mid = multipart_t::create(49.0f);
-        mid.append(std::move(tmp));
-        assert(mid.size() == 2);
-        assert(tmp.empty());
-
-        multipart_t merged(std::move(mid));
-        merged.prepend(std::move(head));
-        merged.append(std::move(tail));
-        assert(merged.size() == 6);
-        assert(head.empty());
-        assert(tail.empty());
-
-        ok = merged.send(output);
-        assert(merged.empty());
-        assert(ok);
-
-        multipart_t received(input);
-        assert(received.size() == 6);
-
-        str = received.popstr();
-        assert(received.size() == 5);
-        assert(str == "One");
-
-        str = received.popstr();
-        assert(received.size() == 4);
-        assert(str == "Two");
-
-        num = received.poptyp<float>();
-        assert(received.size() == 3);
-        assert(num == 49.0f);
-
-        str = received.popstr();
-        assert(received.size() == 2);
-        assert(str == "Fifty");
-
-        str = received.popstr();
-        assert(received.size() == 1);
-        assert(str == "Ninety-nine");
-
-        str = received.popstr();
-        assert(received.empty());
-        assert(str == "One-hundred");
-
-        return 0;
-    }
-
 private:
     // Disable implicit copying (moving is more efficient)
     multipart_t(const multipart_t& other) ZMQ_DELETED_FUNCTION;
     void operator=(const multipart_t& other) ZMQ_DELETED_FUNCTION;
 };  // class multipart_t
 
-#endif  // ZMQ_HAS_RVALUE_REFS
-
 inline std::ostream& operator<<(std::ostream& os, const multipart_t& msg)
 {
     return os << msg.str();
 }
 
+#endif  // ZMQ_HAS_RVALUE_REFS
+
+#if defined(ZMQ_BUILD_DRAFT_API) && defined(ZMQ_CPP11) && defined(ZMQ_HAVE_POLLER)
+    class active_poller_t
+    {
+    public:
+        active_poller_t () = default;
+        ~active_poller_t () = default;
+
+        active_poller_t(const active_poller_t&) = delete;
+        active_poller_t &operator=(const active_poller_t&) = delete;
+
+        active_poller_t(active_poller_t&& src) = default;
+        active_poller_t &operator=(active_poller_t&& src) = default;
+
+        using handler_t = std::function<void(short)>;
+
+        void add (zmq::socket_t &socket, short events, handler_t handler)
+        {
+            auto it = decltype (handlers)::iterator {};
+            auto inserted = bool {};
```

Comment:
depending on default initialization here makes it only harder to read.

Reply:
Hm. I don't think so. Initializing with a value explicitly for which there is no code path where it could ever be read would be confusing IMO.

Reply:
There is not difference in initializing it explicitly or default. But there is readability and maintainability gain. Explicit initialization makes it harder to break things in the future IMO. Your call here.

Reply:
I can understand your point of view regarding future changes. I disagree about readability though. If I read some variable is initialized with a specific value I assume this value will be used in some case, and search for a long time if there is no such case.

Anyway, there is no way without some disadvantage. What is more important is to do this consistently throughout cppzmq IMO (I am not sure if there is any other comparable place at the moment). We should prescribe either way in a brief style guide.

Reply:
Ok. Agree.
We would not have this discussion if we switch poller to C++14. What do you think about that?

Reply:
I don't think this would be a good idea at the moment. I read that people can't use fbzmq e.g. because it requires C++14 support. For our own uses we will stick to VS2015 for several years to come e.g., which has only very limited support for C++14 or even C++17 features. I think there are also many other environments that at least have no recent gcc/clang by default. We already restrict the platforms quite substantially compared to libzmq itself which does not require C++11.